implement RSNT_ENABLE_LMOD_CACHE=auto mode that enables Lmod cache or…

mboisson opened the pull request:
… not depending on logic

This is to attempt to fix https://github.com/ComputeCanada/software-stack/issues/110 until Lmod has better support for that kind of modules.

ofisette commented:
The logic seems alright, but I am confused by `vendor_id` vs `cpu_vendor_id`, especially since it looks like it is not consistent between Gentoo and Nix. Are these two different variables, or perhaps a typo?

I am also confused by the `mode() == "load"` check being done for `auto` mode but not for `yes`.

mboisson commented:
> The logic seems alright, but I am confused by `vendor_id` vs `cpu_vendor_id`, especially since it looks like it is not consistent between Gentoo and Nix. Are these two different variables, or perhaps a typo?

That was just written with a different name between the two modules. 

> 
> I am also confused by the `mode() == "load"` check being done for `auto` mode but not for `yes`.

That's to avoid checking the filesystem for operations other than module load.

ofisette commented:
In `gentoo/2020.lua.core`, you refer to both `vendor_id` and `cpu_vendor_id`. I think in that file `cpu_vendor_id` is a typo or copy/paste error from `nixpkgs`.

I do not understand why not checking for operations other than `module load`, as the issue happens when using `module spider`. But my main point is that you do this check for the `auto` setting and not for the `yes` setting, which seems weird. I would expect this check to either not be done at all, or to be done for both `auto` and `yes`.

mboisson commented:
> In `gentoo/2020.lua.core`, you refer to both `vendor_id` and `cpu_vendor_id`. I think in that file `cpu_vendor_id` is a typo or copy/paste error from `nixpkgs`.

Ah, good point, that was a bad copy-paste.  I renamed them all to `cpu_vendor_id` to avoid the confusion and possible bugs down the road. 

> I do not understand why not checking for operations other than `module load`, as the issue happens when using `module spider`. But my main point is that you do this check for the `auto` setting and not for the `yes` setting, which seems weird. I would expect this check to either not be done at all, or to be done for both `auto` and `yes`.

I understand the interrogation. I modified it to make it clearer.

bartoldeman commented on modules/gentoo/2020.lua.core:
```diff
@@ -60,9 +60,23 @@ elseif arch == "avx" then
+	if os.getenv("RSNT_ENABLE_LMOD_CACHE") == "yes" then
+		append_path("LMOD_RC", "/cvmfs/soft.computecanada.ca/config/lmod/lmodrc/cache_" .. arch .. "_" .. cpu_vendor_id .. ".lua")
+	elseif os.getenv("RSNT_ENABLE_LMOD_CACHE") == "auto" then
+		local has_local_paths = os.getenv("RSNT_LOCAL_MODULEPATHS") or "no"
```
this could be much simplified, and also take advantage of short-circuit boolean logic:

```
if not os.getenv("RSNT_LOCAL_MODULEPATHS") and not posix.stat(pathJoin(home, ".local/easybuild/modules")) then
	append_path("LMOD_RC", "/cvmfs/soft.computecanada.ca/config/lmod/lmodrc/cache_" .. arch .. "_" .. cpu_vendor_id .. ".lua")
end

mboisson replied:
done

bartoldeman commented on modules/nixpkgs/16.09.lua.core:
```diff
@@ -70,8 +70,23 @@ else
+	if os.getenv("RSNT_ENABLE_LMOD_CACHE") == "yes" then
+		append_path("LMOD_RC", "/cvmfs/soft.computecanada.ca/config/lmod/lmodrc/cache_" .. arch .. "_" .. cpu_vendor_id .. ".lua")
+	elseif os.getenv("RSNT_ENABLE_LMOD_CACHE") == "auto" then
+		local has_local_paths = os.getenv("RSNT_LOCAL_MODULEPATHS") or "no"
```
same here!

mboisson replied:
done
